includes/class-newspack-popups-inserter.php

```diff
@@ -180,27 +180,25 @@ public static function insert_popup( $content = '', $popup = [] ) {
 			return $popup_markup . $content;
 		}
 
-		$position  = 0;
-		$positions = [];
-		$close_tag = '</p>';
-		while ( stripos( $content, $close_tag, $position ) !== false ) {
-			$position    = stripos( $content, '</p>', $position ) + strlen( $close_tag );
-			$positions[] = $position;
-		}
-		$total_length       = strlen( $content );
-		$percentage         = intval( $popup['options']['trigger_scroll_progress'] ) / 100;
-		$precise_position   = $total_length * $percentage;
-		$insertion_position = $total_length;
-		foreach ( $positions as $position ) {
-			if ( $position >= $precise_position ) {
-				$insertion_position = $position;
-				break;
+		$percentage       = intval( $popup['options']['trigger_scroll_progress'] ) / 100;
+		$total_length     = strlen( $content );
+		$precise_position = $total_length * $percentage;
+
+		$blocks      = parse_blocks( $content );
+		$pos         = 0;
+		$output      = '';
+		$is_inserted = false;
+		foreach ( $blocks as $block ) {
+			$block_content = render_block( $block );
+			$pos          += strlen( $block_content );
+			if ( ! $is_inserted && $pos >= $precise_position ) {
+				$output     .= '<!-- wp:shortcode -->' . $popup_markup . '<!-- /wp:shortcode -->';
+				$is_inserted = true;
 			}
+			$output .= $block_content;
```

Comment:
A site with this plugin active will have its blocks rendered at priority 1 in `the_content`. I'm not sure whether this will cause issues from other plugins that assume things, but I think there is definitely the potential for some weirdness. For example, a plugin that does something with raw block content and hooks into `the_content` at priority 4 will not have the raw block content it expects. The unknown unknowns here are my main concern. 

An approach that only inserts the shortcode block at the appropriate place and doesn't otherwise modify the content would solve this issue, e.g.
```
		$blocks      = parse_blocks( $content );
		$pos         = 0;
		$output      = '';
		$is_inserted = false;
		foreach ( $blocks as $block ) {
             $original_block_markup = '?????'; // Need to figure this out somehow I think.
			$block_content = render_block( $block );
			$pos          += strlen( $block_content );
			if ( ! $is_inserted && $pos >= $precise_position ) {
				$output     .= '<!-- wp:shortcode -->' . $popup_markup . '<!-- /wp:shortcode -->';
				$is_inserted = true;
			}
			$output .= $original_block_markup; // Let the block parser handle the block rendering at `the_content` priority ~9. 
```

It's a difficult problem to figure out the original block markup from the parsed block, though. I'm not sure how that would work.

Reply:
I'm not that familiar with the potential implications of priority here. Could you please share some code that a plugin might introduce and that would break things?

Reply:
> Could you please share some code that a plugin might introduce and that would break things?

For example, if there were two plugins injecting popups into the content in the exact same way as this PR (Newspack Popups A and Newspack Popups B), only the first plugin would correctly insert the popup. Newspack Popups B wouldn't work correctly.

Here's a contrived example that demonstrates it further:
```
// Remove all image blocks from the content.
add_filter( 'the_content', function( $content ) {
	$regex = '#<!-- wp:image.*\/wp:image -->#sU';
	return preg_replace( $regex, '', $content );
}, 2 );
```

Without this PR, this will correctly remove all image blocks from the content. With this PR active, this will do nothing.

I fear I may be overly concerned with a "correct" logic for this, though. I've been thinking about this further. I think we'll likely be OK. Let's go ahead and release this version and see if anything surfaces. It seems pretty unlikely another plugin would do something similar.

Reply:
Leaving a note here about a better solution I've thought of because I think this did cause issues after all, in the edge case where 2+ inline campaigns are used.

We can combine the different approaches by injecting popups into the **fully rendered** content (run the filter at priority `99` instead of `1`) and **only counting top-level paragraph elements**. The tricky part is how to accurately count top-level paragraphs (it's not feasible with regex). [DOMDocument](https://www.php.net/manual/en/class.domdocument.php) is probably the most straightforward solution I think.